### DISCUSSION TITLE
CoPilot: Privacy & DataMining

### DISCUSSION THREAD

**Discussion by MentalGear** [community/community#7263] [Copilot Conversations] [answered]
Based on the official [CoPilot Privacy Policy](https://docs.github.com/en/github/copilot/about-github-copilot-telemetry),"**source code snippets, and URLs of repositories and file paths**." are collected. [1] Even though Microsoft states that access is strictly confined to employees of theirs, and ["GitHub Copilot does not use these URLs, file paths, or snippets collected in your telemetry as suggestions for other users of GitHub Copilot."](https://docs.github.com/en/github/copilot/github-copilot-telemetry-terms) this could only mean in direct suggestions and it seems entirely probable that code snippets and other sensitive data could end up in a training dataset for the ML NLP (GTP-3) engine, resulting in code snippet suggestions for other users down the line.

This is a huge problem for property software engineers and people in general working with sensitive data, and the following should have been put in place from the start:
1. It should be prominently highlighted in the extension description page that data is collected.
2. Telemetry/DataCollection should be able to be turned off (I didn't find a setting for CoPilot to turn it off - which should be standard).

[1]
> How the data is shared
> 
> The telemetry data is stored securely on GitHub systems, with appropriate encryption in place. We know user edit actions, **source code snippets, and URLs of repositories and file paths are sensitive data**. Consequently, access is strictly controlled. The data can only be accessed by (1) named GitHub personnel (employees and contractors) working on the GitHub Copilot team or on the GitHub platform health team, (2) Microsoft personnel (employees and contractors) working on or with the Azure and/or GitHub Copilot teams, and (3) employees of OpenAI who work on GitHub Copilot.

**Comment by ib0b** [comment]
Was just looking for this.
Same wish it had that feature, just got the technical preview today
most of my projects deal with a lot of sensitive API keys embedded in code since they are mostly short scripts

While copilot is really good, and I understand they need to collect data on snippets in the technical preview to improve the product, another possibility is to include a review feature. An example would be adding an option for the user to cherry-pick, which code snippets will be shared by copilot engineers, while other snippets will only be only seen by the GPT-3 and not stored.

**Reply by mardub1635** [reply]
My question is probably within this scope but might be technically slightly different, not sure if I should post it here.

If inside the working folder there is stored personal information, what happen to this personal information if copilot is active in the folder? This information is not in the code per se. But it is contained in for instance a ".txt" file that the code may open.
Example screenshot:

![code_reading_sensitive_info](https://user-images.githubusercontent.com/13570586/160999948-fcca5b18-b1da-4ff2-a1ac-572501aa4eb9.png)

Is any of the information contained inside the .txt sent to copilot servers?

Let me know if my comment does not belong here and I delete or move it place.

**Reply by Tall1n** [reply]
Since copilot gives you suggestions of functions you defined in other files, I assume they collect everything in the directory.

**Reply by valentinoPereira** [reply]
It's really fishy that microsoft is not releasing any official statement regarding collection of data. They do collect and must admit it, instead of keeping majority of the users in the dark

**Reply by emmenlau** [reply]
> It's really fishy that microsoft is not releasing any official statement regarding collection of data. They do collect and must admit it, instead of keeping majority of the users in the dark

I'm not sure this is correct anymore. I've looked into the difference of paid and free account, and in the process there was a comparsion page that mentioned what code snippets are sent to MS servers, and in which cases they are stored (or not). Maybe check this and see if that answers your question? It was somewhere throughout the free trial account registration: https://github.com/features/copilot/#pricing

**Reply by mardub1635** [reply]
And still no one could answer concretely if in my example ultimately microsoft has collected the info that Miss Doe lives in Tatawin street 5 (see previous message with screenshot).

Maybe my example is not simple enough? However I thought txt stored in working folders is trivial.

**Reply by Fredbay** [reply]
I think your question makes perfect sense and I am actually wondering the exact same thing. Have managed to find an answer to this question else where?

**Reply by mardub1635** [reply]
@Fredbay  Not clearly yet. And honestly, I am unsatisfied of the vagueness and jargoning  responses we get to a question that is simple: 
- Concretely, are people telling us here that if I have a file.txt that contains the information "Miss Doe has HIV" in my current programming folder microsoft/copilot knows it? The answer should be YES or NO.
- do microsoft/copilot read this information in any way? And if yes, do github/copilot send this data anywhere else than on my physical computer? even temporarily?
Thank you for transparency and clarity of future responses. Not all the people using copilot are familiar to notions of clouds, a.i., snippets etc. microsoft awe us a clear layman friendly response with concrete examples.

**Comment by ib0b** [comment, marked as answer]
Seems you now can disable **_snippet collection._**
by **[unchecking Allow GitHub to use my code snippets for product improvements](https://github.com/settings/copilot)**  - is checked by default if you are transitioning from technical preview
<img width="1255" alt="privacy" src="https://user-images.githubusercontent.com/18010106/175777281-66c25e7a-2289-4824-b9f5-c087603ea3d5.png">

**Reply by KaziiTheAvali** [reply]
is there a way for those who do not have github copilot to disable the ability for their code to be used

**Comment by SeidChr** [comment]
There has already been a screenshot, but since i already made it, i thought i'd share it as well (more helps more):

Disable the Data-Collection for Github Copilot Free:

![image](https://github.com/user-attachments/assets/a4b0467c-75db-4e0e-98bd-bf58bc8742e4)
